src/backend/distributed/commands/call.c

```diff
@@ -0,0 +1,128 @@
+#include "distributed/commands.h"
+#include "distributed/connection_management.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/remote_commands.h"
+#include "distributed/shard_pruning.h"
+#include "distributed/worker_manager.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "tcop/dest.h"
+#include "utils/syscache.h"
+
+
+static bool CallFuncExprRemotely(DistObjectCacheEntry *procedure,
+								 FuncExpr *funcExpr, const char *queryString,
+								 DestReceiver *dest);
+
+
+/*
+ * CallDistributedProcedure calls a stored procedure on the worker if possible.
+ */
+bool
+CallDistributedProcedureRemotely(CallStmt *callStmt, const char *queryString,
+								 DestReceiver *dest)
+{
+	DistObjectCacheEntry *procedure = NULL;
+	FuncExpr *funcExpr = callStmt->funcexpr;
+	Oid functionId = funcExpr->funcid;
+
+	procedure = LookupDistObjectCacheEntry(ProcedureRelationId, functionId, 0);
+	if (procedure == NULL)
+	{
+		return false;
+	}
+
+	return CallFuncExprRemotely(procedure, funcExpr, queryString, dest);
+}
+
+
+/*
+ * CallFuncExprRemotely calls a procedure of function on the worker if possible.
+ */
+static bool
+CallFuncExprRemotely(DistObjectCacheEntry *procedure, FuncExpr *funcExpr,
+					 const char *queryString, DestReceiver *dest)
+{
+	Oid colocatedRelationId = InvalidOid;
+	Node *partitionValue = NULL;
+	List *shardList = NIL;
+	ShardInterval *shard = NULL;
+	List *placementList = NIL;
+	ListCell *placementCell = NULL;
+	WorkerNode *preferredWorkerNode = NULL;
+	int connectionFlags = 0;
+	MultiConnection *connection = NULL;
+
+	if (IsMultiStatementTransaction())
+	{
+		ereport(DEBUG2, (errmsg("cannot push down CALL in multi-statement transaction")));
+		return false;
+	}
+
+	colocatedRelationId = ColocatedTableId(procedure->colocationId);
+	if (colocatedRelationId == InvalidOid)
+	{
+		ereport(DEBUG2, (errmsg("stored procedure does not have co-located tables")));
+		return false;
+	}
+
+	partitionValue = list_nth(funcExpr->args, procedure->distributionArgIndex);
+
+	shardList = ShardsForPartitionColumnValue(colocatedRelationId, partitionValue);
+	if (list_length(shardList) != 1)
+	{
+		ereport(DEBUG2, (errmsg("cannot push down CALL when there are overlapping "
+								"shards")));
+		return false;
+	}
+
+	shard = (ShardInterval *) linitial(shardList);
+	placementList = FinalizedShardPlacementList(shard->shardId);
+	if (list_length(placementList) != 1)
+	{
+		/* punt on reference tables for now */
+		ereport(DEBUG2, (errmsg("cannot push down CALL without a distribution column")));
+		return false;
+	}
+
+	foreach(placementCell, placementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		WorkerNode *workerNode = FindWorkerNode(placement->nodeName, placement->nodePort);
+
+		if (workerNode->hasMetadata)
+		{
+			/* we can execute this procedure on the worker! */
+			preferredWorkerNode = workerNode;
```

Comment:
I'd think we'd offer round robin as an option here

Reply:
Would make sense when co-locating with a reference table.